internal/storage/bucket_handle.go

```diff
@@ -96,3 +96,40 @@ func (b *bucketHandle) StatObject(ctx context.Context, req *gcs.StatObjectReques
 
 	return
 }
+
+func (bh *bucketHandle) CreateObject(ctx context.Context, req *gcs.CreateObjectRequest) (o *gcs.Object, err error) {
+	obj := bh.bucket.Object(req.Name)
+
+	if req.GenerationPrecondition != nil && *req.GenerationPrecondition != 0 {
```

Comment:
why not two conditions seperately like it is done for other methods. It avoid nesting of code and improves readability.

Reply:
As I can see in other methods present in this file, request object contains one Generation (not GenerationPrecondition) and one MetagenerationPrecondition. This is the first method, in which request contains both precondition. Also, we can assign one storage.Conditions to obj (using obj.If). If we will write separate if statement for both condition, there might be chance that the later condition assignment will override the previous assignment.

Reply:
1. what is the difference between generation and generation precondition?
2. lets not have might be answers. Please add two conditions seperately and check what is happening. I believe you can check the query in debug mode to see what is generated.

Reply:
Generation is used to select the object on which we want to perform the operation.
While the preconditions is used to make the operation conditional. like, IfGenerationMatch, IfGenerationNotMatch, IfMetaGenerationMatch, IfMetaGenerationNotMatch.

For details: https://cloud.google.com/storage/docs/json_api/v1/objects/patch

Reply:
Create a small taks for this: https://b.corp.google.com/issues/248985083